### PR TITLE
content/blog: Add blog post for GSoC'24: Env Var for Unikraft Builds

### PR DESCRIPTION
**Pull request by UjjwalMahar**
Added the first blog post for GSoC'24 with the progress I have made.

### CODE REVIEW

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 1]
```diff
@@ -0,0 +1,62 @@
+---
+title: "GSoC'24: Supporting User-provided, Long-lived Environmental, Variables for Unikraft Builds"
+description: |
+  The goal of this project is to improve Unikraft's
+  tooling—more especially, the Kraft command-line tool—so that it can dynamically inject
+  user-supplied variables into the build system.
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:6]:
```suggestion
  The goal of this project is to improve Unikraft's tooling, more especially, the `kraft` command-line too, so that it can dynamically inject user-supplied variables into the build system. 
```

No need to add spacing

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 2]
```diff
@@ -0,0 +1,62 @@
+---
+title: "GSoC'24: Supporting User-provided, Long-lived Environmental, Variables for Unikraft Builds"
+description: |
+  The goal of this project is to improve Unikraft's
+  tooling—more especially, the Kraft command-line tool—so that it can dynamically inject
+  user-supplied variables into the build system. 
+publishedDate: 2024-06-19
+tags:
+- GSoC'24   
+- Unikernals
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:10]:
```suggestion
- Unikernels
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+- Unikraft
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:11]:
```suggestion
- KraftKit
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+### Project Overview
+
+Unikraft is a highly adaptable cloud-based library operating
+system with a high degree of modularity. However, developers' ability to add custom variables to the build process is hampered
+by the rigidity of the current tooling. The goal of this project is to improve Unikraft's
+tooling—more especially, the Kraft command-line tool—so that it can dynamically inject
+user-supplied variables into the build system. Features like changing GCC versions and
+compile-time settings are included in this. Developers can customize Unikraft builds to meet
+specific needs by providing dynamic configuration, enhancing flexibility.
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:30]: Newlines on new sentences only please.

```suggestion
Unikraft is a highly adaptable cloud-based library operating system with a high degree of modularity.
However, developers' ability to add custom variables to the build process is hampered by the rigidity of the current tooling.
The goal of this project is to improve Unikraft's tooling—more especially, the Kraft command-line tool—so that it can dynamically inject user-supplied variables into the build system.
Features like changing GCC versions and compile-time settings are included in this.
Developers can customize Unikraft builds to meet specific needs by providing dynamic configuration, enhancing flexibility.
```

Comment (craciunoiuc) [reply content/blog/2024-06-19-gsoc-environment-variables.mdx:30]: Same for other cases

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+- Tooling Framework Improvement
+- Integration with KraftKit Configuration System
+- Modification of Build Invocation Mechanism
+- Default Configuration Setting for CC
+- Kraftfile Integration
+- Dynamic Injection Using Flags in the Command Line
+
+### Progress
+
+Over the past few weeks, I have been following the instructions given by my mentors in the form [Steps](https://github.com/unikraft/kraftkit/issues/673)
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:42]:
```suggestion
Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make Setup and Build:**
+   Configured the development environment and successfully built a sample Unikraft application using Make.
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:46]:
```suggestion
  Configured the development environment and successfully built a sample Unikraft application using Make.
```

Not sure if 3 spaces are intended?

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
+     I have experimented with various GCC versions and compile-time settings by passing different options to the make command, ensuring compatibility and robustness.
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:49]:
```suggestion
  I have experimented with various GCC versions and compile-time settings by passing different options to the make command, ensuring compatibility and robustness.
```

Not sure if 5 spaces are intended?

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+### Next Steps 
+
+For the next steps I aim to get started on the actual implementation of the following - 
+
+1. **Hardcode-Inject Toolchain Option:** Hardcode-inject a toolchain option within the KraftKit make package to evaluate its functionality and determine if redesigning is necessary.
+2. **Global Toolchain Config Variable:** Introduce the Toolchain global config variable to streamline configuration management.
+3. **Kraftfile Toolchain Element:** Introduce the toolchain element in the Kraftfile, drawing suggestions from previous features like environmental variables and readme elements.
+4. **CLI Toolchain Options:** Introduce command-line options/commands to set and list toolchain variables, enhancing user control and flexibility.
+5. **Kraftfile-Level Toolchains:** As a bonus, implement toolchain definitions at the Kraftfile level to provide even more customization.
```
Comment (craciunoiuc) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:59]:
```suggestion
5. **Kraftfile-Level Toolchains:** Implement toolchain definitions at the Kraftfile level to provide even more customization.
```

**Review by craciunoiuc — CHANGES_REQUESTED** [review]
some comments from my side

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+### Next Steps 
+
+For the next steps I aim to get started on the actual implementation of the following - 
+
+1. **Hardcode-Inject Toolchain Option:** Hardcode-inject a toolchain option within the KraftKit make package to evaluate its functionality and determine if redesigning is necessary.
+2. **Global Toolchain Config Variable:** Introduce the Toolchain global config variable to streamline configuration management.
+3. **Kraftfile Toolchain Element:** Introduce the toolchain element in the Kraftfile, drawing suggestions from previous features like environmental variables and readme elements.
+4. **CLI Toolchain Options:** Introduce command-line options/commands to set and list toolchain variables, enhancing user control and flexibility.
+5. **Kraftfile-Level Toolchains:** Implement toolchain definitions at the Kraftfile level to provide even more customization.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:55]:
```suggestion
```
This was slipped on our end when creating the issue. Point 3 and 5 are actually the same thing

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 10]
```diff
@@ -0,0 +1,55 @@
+---
+title: "GSoC'24: Supporting User-provided, Long-lived Environmental, Variables for Unikraft Builds"
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:2]:
```suggestion
title: "GSoC'24: Supporting User-provided, Long-lived Environmental Variables for Unikraft Builds"
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+description: |
+  The goal of this project is to improve Unikraft's tooling, more especially, the `kraft` command-line too, so that it can dynamically inject user-supplied variables into the build system.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:4]:
```suggestion
  The goal of this project is to improve Unikraft's tooling, more especially, the `kraft` command-line tool, so that it can dynamically inject user-supplied variables into the build system. 
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+tags:
+- GSoC'24   
+- Unikernels
+- KraftKit
+- Golang
+authors:
+- Ujjwal Mahar
+---
+
+## GSoC'24: Supporting User-provided, Long-lived Environmental, Variables for Unikraft Builds
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:15]:
```suggestion
## GSoC'24: Supporting User-provided, Long-lived Environmental Variables for Unikraft Builds
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+
+### Project Overview
+
+Unikraft is a highly adaptable cloud-based library operating system with a high degree of modularity.
+However, developers' ability to add custom variables to the build process is hampered by the rigidity of the current tooling.
+The goal of this project is to improve Unikraft's tooling—more especially, the Kraft command-line tool—so that it can dynamically inject user-supplied variables into the build system.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:24]:
```suggestion
The goal of this project is to improve Unikraft's tooling, specifically the `kraft` command-line tool, so that it can dynamically inject user-supplied variables into the build system.
```
Let's be consistent with how we write `kraft`

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make Setup and Build:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
+  I have experimented with various GCC versions and compile-time settings by passing different options to the make command, ensuring compatibility and robustness.
+
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:46]: I would add a short section here explaining the reasoning behind going through this process manually.
Why did you test this? Are you more familiar with the build system? How do you think this helps with the project?

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+
+### Project Overview
+
+Unikraft is a highly adaptable cloud-based library operating system with a high degree of modularity.
+However, developers' ability to add custom variables to the build process is hampered by the rigidity of the current tooling.
+The goal of this project is to improve Unikraft's tooling—more especially, the Kraft command-line tool—so that it can dynamically inject user-supplied variables into the build system.
+Features like changing GCC versions and compile-time settings are included in this.
+Developers can customize Unikraft builds to meet specific needs by providing dynamic configuration, enhancing flexibility.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:26]:
```suggestion
Developers can customize Unikraft builds to meet specific needs by providing dynamic configurations, enhancing flexibility.
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+- Default Configuration Setting for CC
+- Kraftfile Integration
+- Dynamic Injection Using Flags in the Command Line
+
+### Progress
+
+Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make Setup and Build:**
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:41]: This needs to be reworded, I'm not sure what you mean by `Setup and Build`

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:44]: Same as above, needs to be reworded.

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+- Kraftfile Integration
+- Dynamic Injection Using Flags in the Command Line
+
+### Progress
+
+Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make Setup and Build:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:42]: Briefly mention that the underlying Unikraft build system is Make-based, which is what `kraft` calls into

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
+  I have experimented with various GCC versions and compile-time settings by passing different options to the make command, ensuring compatibility and robustness.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:45]: I'm not sure `compatibility and robustness` is something your experiment was about. 
Sure, the build system is reliable, but this experiment's goal was rather to give you an idea of how the build system behaves when you inject those options, so that you know what injecting them with kraftkit should behave like.
I would reword this to emphasize why you did this and how it's helpful.

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+### Next Steps 
+
+For the next steps I aim to get started on the actual implementation of the following - 
+
+1. **Hardcode-Inject Toolchain Option:** Hardcode-inject a toolchain option within the KraftKit make package to evaluate its functionality and determine if redesigning is necessary.
+2. **Global Toolchain Config Variable:** Introduce the Toolchain global config variable to streamline configuration management.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:52]:
```suggestion
2. **Global Toolchain Config Variable:** Introduce the Toolchain global Kraftkit config variable to streamline configuration management.
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+3. **Kraftfile Toolchain Element:** Introduce the toolchain element in the Kraftfile, drawing suggestions from previous features like environmental variables and readme elements.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:53]: Every other point is formulated as "action: result", so this one stands out to me.
It is also fine as is, I don't have a strong opinion, but I also think you could replace everything after the comma with the benefit of implementing this:
```suggestion
3. **Kraftfile Toolchain Element:** Introduce the toolchain element in the Kraftfile, ... allow per application toolchain options ... .
```
Something along those lines

**Review by LucaSeri — CHANGES_REQUESTED** [review]
Hi,

See the comments from my side. Please also squash your commits.

File: content/blog/2024-06-19-gsoc-environment-variables.mdx (hunk 21 above)
Comment (craciunoiuc) [reply content/blog/2024-06-19-gsoc-environment-variables.mdx]: @UjjwalMahar the idea was to write that part yourself, hence the triple dots, not paste with the dots.

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+  It was necessary to have a good understanding of how kraftkit builds and also a necessary step for 
+  [hacking](https://unikraft.org/docs/cli/hacking) in kraftkit.
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
+  I have experimented with various GCC versions and compile-time settings by passing different options to the make command.
+  This gave me a better understanding of how to utilize different toolchain options available in kraftkit and would eventually
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:48]:
```suggestion
  This gave me a better understanding of how to utilize different toolchain options in Unikraft and would eventually 
```

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+1. **Test Building an Application via Make:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+  It was necessary to have a good understanding of how kraftkit builds and also a necessary step for 
+  [hacking](https://unikraft.org/docs/cli/hacking) in kraftkit.
+
+2. **Exploring Different Toolchain Options Toolchain Testing:**
+  I have experimented with various GCC versions and compile-time settings by passing different options to the make command.
+  This gave me a better understanding of how to utilize different toolchain options available in kraftkit and would eventually 
+  it would help me with my project.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:49]: Please use one line per sentence and one sentence per line

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+
+### Progress
+
+Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+  It was necessary to have a good understanding of how kraftkit builds and also a necessary step for 
+  [hacking](https://unikraft.org/docs/cli/hacking) in kraftkit.
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:44]: Please use one line per sentence and one sentence per line

File: content/blog/2024-06-19-gsoc-environment-variables.mdx [hunk 25]
```diff
@@ -0,0 +1,58 @@
...
+- Dynamic Injection Using Flags in the Command Line
+
+### Progress
+
+Over the past few weeks, I have been following the instructions given by my mentors in the issue [Steps](https://github.com/unikraft/kraftkit/issues/673)
+I have completed two of the steps which include:
+
+1. **Test Building an Application via Make:**
+  Configured the development environment and successfully built a sample Unikraft application using Make.
+  It was necessary to have a good understanding of how kraftkit builds and also a necessary step for
```
Comment (LucaSeri) [inline content/blog/2024-06-19-gsoc-environment-variables.mdx:43]:
```suggestion
  It was necessary to have a good understanding of how Unikraft builds and also a necessary step for 
```

**Review by LucaSeri — CHANGES_REQUESTED** [review]

**Review by craciunoiuc — APPROVED** [review]
All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>
Approved-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

**Comment by craciunoiuc** [comment]
There are some linter errors left, but the content looks overall fine

<https://github.com/unikraft/docs/actions/runs/9598747477/job/26477280158?pr=431>

**Comment by StefanJum** [comment]
@LucaSeri is this ok with you?

@UjjwalMahar please reword the commit message to fit to 75-80 characters.

**Review by LucaSeri — APPROVED** [review]
Looks good, thanks!

Reviewed-by: Luca Seritan luca.seritan@gmail.com
Approved-by: Luca Seritan luca.seritan@gmail.com

**Comment by StefanJum** [comment]
@UjjwalMahar the commit message still overflows the GitHub render limit (it renders as `content/blog: Add blog post for GSoC'24:Environmental Variables for U…`. You can shorten it a bit further.
